docs/adr/0003-graphql-layer-between-backend-frontend.md

```diff
@@ -0,0 +1,98 @@
+# GraphQL layer between backend and frontend
+
+Date: 2020-04-30
+
+## Status
+
+**Accepted**
+
+## Related tickets (not available for public)
+
+SIG-2568
+
+## Context
+
+Since our current backend only serves data through REST and there is no possibility to generate
+types and or interfaces, there is no also no automated way to validate data on the frontend
+which is coming from the backend.
+
+Besides of that there are too many inconstencies between different backend calls and the current
+backend API documentation is not always correct and/or up to date because it's written manually
+and not generated from code.
+
+Because of this our `React components` are overcomplicated since we need to transform the API output to a
+format that the API can consume and that leads to bloat and too much client-side logic.
+
+## Proposal
+
+We would like to introduce a thin layer between the backend and our frontend which should expose
+backend data through GraphQL Queries and should expose backend PUT/POST/DELETE actions via GraphQL
+mutations.
```

Comment:
I think that introducing a "thin layer" between the frontend and backend does actually solve the problems, but only moves the problem to the "thin layer". I am not sure how this helps anyone.

I think it would be better to solve the problems of the inconsistent API specification in the backend and/or implement a GraphQL interface directly in the backend with [Graphene](https://docs.graphene-python.org/projects/django/en/latest/) or [Ariadne](https://ariadnegraphql.org/).

Reply:
The `thin layer` will not solve `problems` on the frontend, however it will give us a much better `separation of concerns` which should lead to simpler and cleaner code.

It will also allow us to replace quite some custom code/logic with code/logic which is provided and maintained by the GraphQL communities, think for example about using Axios or Fetch, less selectors, all data transforming could find place in the layer, we could reuse the response objects to make mutations, etc. and we can can organize our queries and mutations in one place.

We will also get a decent amount of 'tooling' for free which will make our life easier (Think about query/mutation validation in your editor, one place where you could test or build all queries and mutations in an easy and simple way, great type/interface integration, tab completion on queries and mutations and much more.

You are correct that REST does not give us the flexibility we would like. *Graphene* or *Ariadne* would also be a `solution`. I have talked about GraphQL with the backend team a couple of times, from what I understood, it is not possible to implement it on short term because of various reasons, I think we could implement this relatively fast with this proposed ADR.

Reply:
As the idea is to improve the quality of the frontend code, did you consider refactoring the code within the frontend codebase instead of introducing a new service?

Reply:
> I really like the idea of strengthening the contract between frontend and backend and use tools to automatically generate types and interfaces in the frontend.
> 
> In my opinion it is very important that this contract is implemented end-to-end and not only between the frontend and a new middleware, otherwise we will probably make things worse.

Agree, but from what I understood this would be hard to achieve this (and would take lots of time) in our current backend because of some design decisions in the past. 

> I think with the current technology in place we can already strengthen the contract and automatically generate types and interfaces.

I am open for that and could even use them in this proposed `thin layer`, still they won't give us the power and simplicity of GraphQL.

> If we choose a new technlogy: GraphQL is a great candidate but we can also consider alternatives such as [gRPC(-web)](https://github.com/grpc/grpc-web).

I have worked with `gRPC` before on a extreme large scale, while it is beautiful for what it is designed for I don't think you can compare this to `GraphQL`.

GraphQL is a specification which could theoretically be used over `gRPC` as far as I know some people even started working on this. But without going to far out of scope, I would compare `gRPC` to `websockets` which is something supported by any major browser and GraphQL works through websockets and are 'normally' used for GraphQL subscriptions.

Reply:
@jpoppe we still need to talk about a solution. Before we can talk about solutions we need to know what issues are the FE team is facing. 

Also you mentioned GraphQL a couple of times, and this was/is not a main priority for the team. It is still to be determined if GraphQL will be the solution or that this is a nice to have in the future.

Reply:
Did you consider refactoring the code within the current frontend codebase and add some layering there, instead of introducing a new middleware?

We are currently refactoring some code as well in the backend with Signalen/backend#7. This refactoring makes it easier to introduce a GraphQL interface in the backend.

Adding a new middleware between frontend and backend will make debugging much harder, especially without the required logging and tracing tooling in place.